rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,377 @@
+
+##### 2. Share Link
+
+Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
+by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
+check the status of the upload
+
+##### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct
+links to the files and often need to be parsed into the correct shape for downloading. The correct URL will be generated
+as part of the sanitization process. The backend that runs this should also be given minimal privileges to further protect from
+server-side request forgery.
+
+###### How to download from a shared link
+
+- Dropbox <https://help.dropbox.com/files-folders/share/force-download>
+- Google Drive <https://www.labnol.org/internet/direct-links-for-google-drive/28356/>
+
+##### 4. _Optional_ Queue
+
+This part is optional and depends on the load expected and server type selected. The queue stores the links that need to
+be processed. It buffers the downstream servers from download requests. This allows us to spin up servers as needed.
+It also gives us the ability to retry a download if one of the servers fails to download or upload the file.
+
+###### Why might the download fail
+
+- The downloading server ran out of storage.
+- The downloading server crashed.
+- The CSP experience and outage
+- The user has reached their daily bandwidth allowance from their CSP. [more info](https://help.dropbox.com/files-folders/share/banned-links)
+
+##### 5. Start EC2 Instance to Download File
+
+Spin up EC2 instances with enough storage to download the whole file. TODO: See if spot instances could be used.
+
+##### 6. Download the File
+
+The most time will be spent here downloading the file. Most CSPs offer a way to validate the integrity of your download
+by providing a hash of the data. This hash can be compared with the hash of the downloaded file. If the hash does not match,
+then reattempt the download.
+
+##### 7. Upload to S3
+
+If the file is >5GB in size, upload the file directly to s3 without writing to disk. For files <5GB, the file must
+be completely downloaded before it can be uploaded to s3. The AWS SDK will be used to upload large files in parts.
+Once the upload is complete, the download job will be removed from the queue.
+
+Before uploading calculate the expected hash. Use the [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)
+during a PutObject request to verify the integrity of each chunk.
+
+> QUESTION? Do we need to tell the user or the DB when the file is done do uploading? This makes me think we need an endpoint
+> to finalize the upload. The endpoint will look for the expected file, and if it's found, update the database with the file
```

Comment:
I'm not sure an endpoint is needed. In various steps of the uploading/validation process, it seems like the database should get updated such that at any point in time, the frontend could call `GET /v1/submission/{project_uuid}` and get the full details of their submission including the status on each of the datasets they tried to upload.

But yes, you do need to update the DB :)